pkg/cache/priority_queue.go

```diff
@@ -0,0 +1,191 @@
+package cache
+
+import (
+	"container/heap"
+	"math"
+)
+
+const (
+	defaultCapacity = 8096
+)
+
+// PriorityQueue queue has priority, the first element has the highest priority
+type PriorityQueue struct {
+	queue    *priorityHeap
+	items    map[interface{}]*Entry
+	capacity int
+}
+
+// NewPriorityQueue construct of priority queue
+// -1: unlimited
+// 0: default capacity
+func NewPriorityQueue(capacity int) *PriorityQueue {
+	if capacity < 0 {
+		capacity = math.MaxInt16
+	} else if capacity == 0 {
+		capacity = defaultCapacity
+	}
+
+	return &PriorityQueue{
+		queue:    &priorityHeap{},
+		items:    make(map[interface{}]*Entry),
+		capacity: capacity,
+	}
+}
+
+// Push push element into pq with priority
+// it will refused if used > capacity
+func (pq *PriorityQueue) Push(priority int, x interface{}) bool {
+	v, ok := pq.items[x]
+	if !ok {
+		if pq.Size() >= pq.capacity {
+			return false
+		}
+		v = &Entry{Value: x, Priority: priority}
+		pq.queue.Push(v)
+	} else {
+		pq.Update(v, priority)
+	}
+	pq.items[x] = v
+	return true
+}
+
+// Pop pop the high Priority element from queue
+func (pq *PriorityQueue) Pop() *Entry {
+	entry, ok := pq.queue.Pop().(*Entry)
+	if !ok || entry == nil {
+		return nil
+	}
+	delete(pq.items, entry.Value)
+	return entry
+}
+
+// Peek return the highest priority element
+func (pq *PriorityQueue) Peek() *Entry {
+	return pq.queue.Peek()
+}
+
+// Update update the priority of the given element
+func (pq *PriorityQueue) Update(old *Entry, priority int) {
+	pq.queue.update(old, old.Value, priority)
+}
+
+// GetAll return limited Entries, if limit = -1 ,it will return all entries
+func (pq *PriorityQueue) GetAll() []*Entry {
+	result := pq.queue.GetAll()
+	return result
+}
+
+// RemoveValue remove element
+func (pq *PriorityQueue) RemoveValue(value interface{}) {
+	if entry, ok := pq.items[value]; ok {
+		pq.queue.remove(entry.Index)
+		delete(pq.items, value)
+	}
+}
+
+// RemoveValues remove elements
+func (pq *PriorityQueue) RemoveValues(values []interface{}) {
+	for _, v := range values {
+		pq.RemoveValue(v)
+	}
+}
+
+// Has it will return true if queue has the value
+func (pq *PriorityQueue) Has(v interface{}) bool {
+	if value, ok := pq.items[v]; ok && value.Index != -1 {
+		return true
+	}
+	return false
+}
+
+// Size return the size of queue
+func (pq *PriorityQueue) Size() int {
+	return pq.queue.Len()
+}
+
+// Entry internal struct for record element
+type Entry struct {
+	Value    interface{}
+	Priority int
+	Index    int
```

Comment:
```suggestion
	index    int
```
The value cannot be accessed and modified externally.